File: Ruby/test-doukaku.rb
# example:
#   $ ruby -I . test-doukaku.rb sample-solver.rb sample-data.txt

solver = ARGV.shift
data_file = ARGV.shift

require "#{solver}"

def test(input, expected)
  actual = solve(input)
  if actual == expected
    print '.'
  else
    print <<-EOS

input:     #{input}
expectied: #{expected}
actual:    #{actual}
    EOS
  end
end

File.open(data_file).each_line do |line|
  test(*line.split)
end
puts

File: C/test-doukaku.c
#include <stdio.h>
#include <string.h>
#include "solver.h"

#define STRING_LENGTH 128

void test(const char input[], const char expected[], FILE *fp) {
    char actual[STRING_LENGTH];
    solve(input, actual);
    if(strcmp(actual, expected) == 0) {
        fprintf(fp, ".");
    } else {
        fprintf(
            fp,
            "\n"
            "input:    %s\n"
            "expected: %s\n"
            "actual:   %s\n",
            input,
            expected,
            actual
        );
    }
}

int main(int argc, char *argv[]) {
    char line[STRING_LENGTH];
    FILE *fp = fopen(argv[1], "r");

    while(fgets(line, sizeof(line), fp)) {
        char *p = strchr(line, '\n');
        *p = '\0';
        p = strchr(line, ' ');
        *p++ = '\0';
        test(line, p, stdout);
    }
    printf("\n");

    return 0;
}
